watertap/core/zero_order_sido_reactive.py

```diff
@@ -0,0 +1,416 @@
+
+    # Load reaction identifiers from database
+    dbparams = self.config.database.get_unit_operation_parameters(
+        self._tech_type,
+        subtype=self.config.process_subtype)
+    try:
+        rxn_ids = list(dbparams["reactions"].keys())
+    except KeyError:
+        raise KeyError(
+            f"{self.name} - database provided does not contain a list of "
+            f"reactions for this technology.")
+    # Create indexing set for reactions
+    self.reaction_set = Set(initialize=rxn_ids)
+
+    # Add performance variables
+    self.recovery_frac_mass_H2O = Var(
+        self.flowsheet().time,
+        initialize=0.8,
+        domain=NonNegativeReals,
+        units=pyunits.dimensionless,
+        bounds=(1E-8, 1.0000001),
+        doc='Mass recovery fraction of water in the treated stream')
+    self.removal_frac_mass_solute = Var(
+        self.flowsheet().time,
+        self.config.property_package.solute_set,
+        domain=NonNegativeReals,
+        initialize=0.01,
+        units=pyunits.dimensionless,
+        doc='Solute removal fraction on a mass basis')
+
+    self.reaction_conversion = Var(
+        self.flowsheet().time,
+        self.reaction_set,
+        initialize=0.5,
+        units=pyunits.dimensionless,
+        doc='Reaction conversion based on key reactant')
+
+    # Load reaction conversions
+    for (t, r), v in self.reaction_conversion.items():
+        try:
+            conv = dbparams["reactions"][r]["conversion"]
+        except KeyError:
+            raise KeyError(
+                f"{self.name} - database provided does not "
+                f"contain an entry for conversion for reaction {r}.")
+        v.fix(conv)
+
+    # Intermediate variables and parameters
+    self.extent_of_reaction = Var(
+        self.flowsheet().time,
+        self.reaction_set,
+        initialize=0,
+        units=pyunits.kg/pyunits.s,
+        doc='Reaction extent on a mass basis based on key reactant')
+
+    # Reaction extent equation
+    @self.Constraint(self.flowsheet().time,
+                     self.reaction_set,
+                     doc='Calcuation of reaction extent from conversion')
+    def reaction_extent_equation(b, t, r):
+        # Get key reactant from database
+        try:
+            key_reactant = dbparams["reactions"][r]["key_reactant"]
+        except KeyError:
+            raise KeyError(
+                f"{self.name} - database provided does not "
+                f"contain an entry for key_reactant for reaction {r}.")
+        if key_reactant not in b.properties_in[0].component_list:
+            raise ValueError(
+                f"{self.name} - key_reactant {key_reactant} for reaction {r} "
+                f"is not in the component list used by the assigned property "
+                f"package.")
+        return (b.extent_of_reaction[t, r] ==
+                b.reaction_conversion[t, r] *
+                b.properties_in[t].flow_mass_comp[key_reactant])
```

Comment:
Isn't this not technically the extent_of_reaction? This is the amount of the key reagent reacted. I like how the model is formulated, but I think we should rename the variable to not cause confusion.

Reply:
In another comment, I suggest creating a variable for the generation term, we could just use that with the key reagent index instead of this variable.

Reply:
I was a little confused by this as well. Following the definition of extent of reaction, represented by delta_n/stoich #, and working out how you formulated things, I think the extent of reaction should be  
`extent_of_reaction == reaction_conversion * flow_mass_comp[key_reactant] * MW[key_reactant]` which should work out to be equivalent to moles/s generated divided by stoichiometric number.

Reply:
The problem is that we will not always have a MW for the components (we explicitly allow for empirical mass conversion ratios), hence we can't do this on a molar basis. Thus, this is a mass-based extent of reaction instead.